nixos/tests/zram-generator.nix

```diff
@@ -0,0 +1,17 @@
+import ./make-test-python.nix {
+  name = "zram-generator";
+
+  nodes.machine = { pkgs, ... }: {
+    environment.etc."systemd/zram-generator.conf".text = ''
+      [zram0]
+      zram-size = ram / 2
+    '';
+    systemd.packages = [ pkgs.zram-generator ];
+  };
+
+  testScript = ''
```

Comment:
The test fails for me:

```
(finished: cleanup, in 0.13 seconds)
Traceback (most recent call last):
  File "/nix/store/gyqlgr5ab1zggk714764b6iiy1dzxf5j-nixos-test-driver-1.1/bin/.nixos-test-driver-wrapped", line 9, in <module>
    sys.exit(main())
  File "/nix/store/gyqlgr5ab1zggk714764b6iiy1dzxf5j-nixos-test-driver-1.1/lib/python3.10/site-packages/test_driver/__init__.py", line 112, in main
    driver.run_tests()
  File "/nix/store/gyqlgr5ab1zggk714764b6iiy1dzxf5j-nixos-test-driver-1.1/lib/python3.10/site-packages/test_driver/driver.py", line 142, in run_tests
    self.test_script()
  File "/nix/store/gyqlgr5ab1zggk714764b6iiy1dzxf5j-nixos-test-driver-1.1/lib/python3.10/site-packages/test_driver/driver.py", line 138, in test_script
    exec(self.tests, symbols, None)
  File "<string>", line 1, in <module>
  File "/nix/store/gyqlgr5ab1zggk714764b6iiy1dzxf5j-nixos-test-driver-1.1/lib/python3.10/site-packages/test_driver/machine.py", line 458, in wait_for_unit
    retry(check_active, timeout)
  File "/nix/store/gyqlgr5ab1zggk714764b6iiy1dzxf5j-nixos-test-driver-1.1/lib/python3.10/site-packages/test_driver/machine.py", line 126, in retry
    if fn(False):
  File "/nix/store/gyqlgr5ab1zggk714764b6iiy1dzxf5j-nixos-test-driver-1.1/lib/python3.10/site-packages/test_driver/machine.py", line 441, in check_active
    raise Exception(f'unit "{unit}" reached state "{state}"')
Exception: unit "systemd-zram-setup@zram0.service" reached state "failed"
kill vlan (pid 7)
error: builder for '/nix/store/b5528s695mp1d930gw5laxzw1lwh4n58-vm-test-run-zram-generator.drv' failed with exit code 1;
       last 10 log lines:
       >     exec(self.tests, symbols, None)
       >   File "<string>", line 1, in <module>
       >   File "/nix/store/gyqlgr5ab1zggk714764b6iiy1dzxf5j-nixos-test-driver-1.1/lib/python3.10/site-packages/test_driver/machine.py", line 458, in wait_for_unit
       >     retry(check_active, timeout)
       >   File "/nix/store/gyqlgr5ab1zggk714764b6iiy1dzxf5j-nixos-test-driver-1.1/lib/python3.10/site-packages/test_driver/machine.py", line 126, in retry
       >     if fn(False):
       >   File "/nix/store/gyqlgr5ab1zggk714764b6iiy1dzxf5j-nixos-test-driver-1.1/lib/python3.10/site-packages/test_driver/machine.py", line 441, in check_active
       >     raise Exception(f'unit "{unit}" reached state "{state}"')
       > Exception: unit "systemd-zram-setup@zram0.service" reached state "failed"
       > kill vlan (pid 7)
       For full logs, run 'nix log /nix/store/b5528s695mp1d930gw5laxzw1lwh4n58-vm-test-run-zram-generator.drv'.
```

Reply:
The test is failing as expected, and that's the reason why this hasn't been merged: https://github.com/NixOS/nixpkgs/pull/210190#issuecomment-1378663461. We can incorporate the workaround then call it a day, or we can really fix this on systemd side.

Reply:
Ah sorry. I think this already popped up once or twice. The check for this is in systemd `src/shared/mkfs-util.c`. It invokes `r = find_executable("mkswap", &mkfs);`

In the past, some of these were hardcoded, we used to patch these paths directly (0a41d69968cc6fe19493ec9f06c166066bcfcd45).

These days, most of this is discovered from `$PATH`. We have `0017-inherit-systemd-environment-when-calling-generators.patch`, which together with `systemd.managerEnvironment.PATH` set to `lib.makeBinPath (config.system.fsPackages ++ [cfg.package.util-linux])` should work?

I'm not sure, maybe this regressed during a3e0698bf6fd676d96b89bdb4cd54c73ea502746? @dasJ, do we even need/want that patch anymore?

Independent of that, it looks like there were also attempts to append this to the path of the `systemd-mkswap@.service`/ `systemd-makefs@.service` overrides: https://github.com/NixOS/nixpkgs/pull/120015/files#diff-50e7a8c943eedf4d488db6a69f70cea2c510359e610ec2e8479dba478d7b5011R316, which might be cleaner? (cc @ElvishJerricco)

Reply:
Thanks for your clarification, so adding things to PATH is currently the right way.

Reply:
Let's try to get some agreement on how we want to handle these things for `systemd-mkswap@.service`/ `systemd-makefs@.service`, maybe you can plug into the same thing then.

Reply:
I agree with the general idea of adding per unit overrides. After all reducing global state has always been the point of NixOS. Just that this can be easily forgotten and cause some confusion (like I has in this one) in future PRs.

Reply:
We should probably do it the same way we do `systemd-fsck@.service`

https://github.com/NixOS/nixpkgs/blob/a743cf65d36af9633fe758092e54b0ba4f1f43e7/nixos/modules/system/boot/systemd.nix#L588

Reply:
Btw, `managerEnvironment` is only for PID 1, and isn't propagated to units. `DefaultEnvironment` would do that, which is what systemd stage 1 does to enable `makefs`, but that's obviously not suitable for stage 2.

And @flokli the old PR you linked to was only doing it for stage 1. In the final PR, as I said, it ended up in `DefaultEnvironment`.

Reply:
Ok, so this means this should get a small nixos module that does set it for us?

Reply:
@flokli I would say it should just go in the `systemd` module next to the `fsck` part, since it's really a systemd feature that's being fixed.

Actually, I forgot I kinda did part of this in a toy commit a while back. Supporting `makefs`/`mkswap` is a touch more involved since those units are created by generators. Here's a commit I toyed around with a while ago that switched `autoFormat` and `autoResize` to use `makefs` and `growfs`: https://github.com/NixOS/nixpkgs/commit/bb0cd5d48ae7f4feb85848d7c42aa42185306ae3

The relevant bit is the `systemd` module, where I added a section of fixups for `makefs` (I didn't do `mkswap`). I had to add `overrideStrategy = "asDropin"` because of the generator thing, and I *should* have set `path` the same way its set for `fsck` for consistency but it's a little different

Reply:
Isn't `config.system.path` like containing everything, far beyond the limited set required by `systemd-fsck`. `config.system.fsPackages` is a more reasonable choice.

Reply:
@flokli Oh wait there needs to be a `systemd-zram-setup@` override? I thought we just needed a `systemd-mkswap@` override, which would be fixing a systemd feature.

Reply:
I digged a bit into the source code. https://github.com/systemd/zram-generator/blob/5809ec853930b2ea948a18ef2407e38a585c09e2/src/setup.rs does read a `SYSTEMD_UTIL_DIR` env var set during build, and bakes that path into the binary. It uses this to call out to the `systemd-makefs` binary directly.

This gets set by the `Makefile` (by asking pkgconfig), which we seem to be invoking, and it ends up with the correct path to

`systemd-makefs` itself discovers `mkswap` from `$PATH`, and because `zram-generator` invokes the binary directly, not via the `systemd-mkswap@.service`, overriding the `systemd-mkswap@.service` won't help here. Looks like they don't want to change it either, because not all distros ship with a `systemd-mkswap@.service`: https://github.com/systemd/zram-generator/issues/27

Now our VM test does add `mkswap` to $PATH in an override for `systemd-zram-setup@.service`, so the `systemd-mkswap` binary invoked by `zram-generator` is able to find `mkswap`.

I think we have no choice than to override `systemd-zram-setup@.service` to fix this - either by post-massaging the `.service` file post-build, or by adding an override of that via a NixOS module. I'd personally favor a module, which could also provide a `toINI` translation of the config file.

Reply:
Yea I think a module for this makes sense then, with an option to enable the generator (and configure its generated units with a functioning `path`) and set the config.

Reply:
The module is here: https://github.com/NixOS/nixpkgs/blob/nixos-unstable/nixos/modules/config/zram.nix, just need to be rewritten to take advantage of zram-generator.

Reply:
So should we merge this without a module and migrate that module in a separate PR? Or do both things in this one?

Reply:
Without the module, IMO.